manifests/init.pp

```diff
@@ -210,68 +232,100 @@
-  Enum['yes','no'] $do_not_log_tcpwrappers = $snmp::params::do_not_log_tcpwrappers,
-  Array[String[1]] $trap_handlers = $snmp::params::trap_handlers,
-  Array[String[1]] $trap_forwards = $snmp::params::trap_forwards,
-  Array[String] $snmptrapd_config = $snmp::params::snmptrapd_config,
-  Boolean $manage_client        = $snmp::params::manage_client,
-  $snmp_config                  = $snmp::params::snmp_config,
-  Boolean $autoupgrade          = $snmp::params::autoupgrade,
-  $package_name                 = $snmp::params::package_name,
-  $snmptrapd_package_name       = $snmp::params::snmptrapd_package_name,
-  $snmpd_options                = $snmp::params::snmpd_options,
-  $service_config_perms         = $snmp::params::service_config_perms,
-  $service_config_dir_group     = $snmp::params::service_config_dir_group,
-  Stdlib::Ensure::Service $service_ensure = $snmp::params::service_ensure,
-  $service_name                 = $snmp::params::service_name,
-  Boolean $service_enable       = $snmp::params::service_enable,
-  Boolean $service_hasstatus    = $snmp::params::service_hasstatus,
-  Boolean $service_hasrestart   = $snmp::params::service_hasrestart,
-  $snmptrapd_options            = $snmp::params::snmptrapd_options,
-  Stdlib::Ensure::Service $trap_service_ensure = $snmp::params::trap_service_ensure,
-  $trap_service_name            = $snmp::params::trap_service_name,
-  $trap_service_enable          = $snmp::params::trap_service_enable,
-  $trap_service_hasstatus       = $snmp::params::trap_service_hasstatus,
-  $trap_service_hasrestart      = $snmp::params::trap_service_hasrestart,
-  String[1] $template_snmpd_conf          = $snmp::params::template_snmpd_conf,
-  String[1] $template_snmpd_sysconfig     = $snmp::params::template_snmpd_sysconfig,
-  String[1] $template_snmptrapd           = $snmp::params::template_snmptrapd,
-  String[1] $template_snmptrapd_sysconfig = $snmp::params::template_snmptrapd_sysconfig,
-  Boolean $openmanage_enable    = $snmp::params::openmanage_enable,
-  Boolean $master               = $snmp::params::master,
-  $agentx_perms                 = $snmp::params::agentx_perms,
-  $agentx_ping_interval         = $snmp::params::agentx_ping_interval,
-  $agentx_socket                = $snmp::params::agentx_socket,
-  Integer[0] $agentx_timeout    = $snmp::params::agentx_timeout,
-  Integer[0] $agentx_retries    = $snmp::params::agentx_retries,
-  Boolean $snmpv2_enable        = $snmp::params::snmpv2_enable,
-) inherits snmp::params {
+  Enum['present','absent']                                        $ensure        = 'present',
+  Array[String[1]]                                                $agentaddress  = [ 'udp:127.0.0.1:161', 'udp6:[::1]:161' ],
+  Array[String[1]]                                                $snmptrapdaddr = [ 'udp:127.0.0.1:162', 'udp6:[::1]:162' ],
+  Variant[Undef, String[1], Array[String[1]]]                     $ro_community  = 'public',
```

Comment:
does undef make sense here? As far as I know, you can pass `~` as a parameter, but than puppet will fallback to the default data, `'public'` in this case. IMO the default needs to undef, or remove undef from the Variant.

Reply:
I did that to preserve existing functionality.

See these tests where `ro_community` is an empty array.

https://github.com/voxpupuli/puppet-snmp/blob/021c1b27e68481a99b39cccccb40bdaae1d06889/spec/acceptance/snmp_spec.rb#L36

Reply:
I agree that if the default isn’t `undef`, allowing `undef` is almost always wrong.

@bastelfreak using `~` in hiera would be the *only* way to override it to `undef`. You couldn’t do this via a class declaration though (where it’d be interpreted as use the default).  What does the code do if it’s `undef` vs empty array?

Reply:
`[*@ro_community].compact.each` eew.

Splat operator, does that return anything for `nil`?? I guess so otherwise the `rw_community` default wouldn’t work.

So some users *might* be using hiera to set `undef`, but people declaring the class have to use an empty array.

Why don’t we remove all the mess and inconsistency by only allowing arrays, (including empty arrays)? This is a breaking change, but this is for a new major release anyway.

@ghoneycutt as you’ve seen in https://github.com/voxpupuli/puppet-snmp/issues/170 removing all of this functionality has been threatened for a while. The readme actually says we’ll do this in 5.0. So let’s officially undeprecate traditional access control, but also simplify the code and template in the process as a compromise?

Reply:
I just wanted to do this piece of work to give back to the community and modernize the module a bit without changing the tests. I've gone as far as I'd like to with this module. Feel free to base your work on top of mine with additional commits.

Reply:
That’s fair enough and your improvements to this module are greatly appreciated.

I do think @bastelfreak was ok to point out allowing `undef` whilst defaulting to a value is possibly suspect and worth investigating. If worth doing, the simplification I suggested could always be done in a separate PR, (but before the 5.0 release).

Reply:
#170 is a separate issue. It would seem we need to keep the old functionality. Anyhow, let's keep that discussion over there.

Reply:
BTW, `undef` for `ro_community` does seem wrong. I was purposely coding against the tests that were already present as opposed to making feature enhancements.